crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs

```diff
@@ -63,23 +63,42 @@ impl<'a> FunctionContext<'a> {
     /// The returned parameter type list will be flattened, so any struct parameters will
     /// be returned as one entry for each field (recursively).
     fn add_parameters_to_scope(&mut self, parameters: &Parameters) {
-        for (id, _, _, typ) in parameters {
-            self.add_parameter_to_scope(*id, typ);
+        for (id, mutable, _, typ) in parameters {
+            self.add_parameter_to_scope(*id, typ, *mutable);
         }
     }
 
     /// Adds a "single" parameter to scope.
     ///
     /// Single is in quotes here because in the case of tuple parameters, the tuple is flattened
     /// into a new parameter for each field recursively.
-    fn add_parameter_to_scope(&mut self, parameter_id: LocalId, parameter_type: &ast::Type) {
+    fn add_parameter_to_scope(
+        &mut self,
+        parameter_id: LocalId,
+        parameter_type: &ast::Type,
+        mutable: bool,
+    ) {
         // Add a separate parameter for each field type in 'parameter_type'
-        let parameter_value =
-            Self::map_type(parameter_type, |typ| self.builder.add_parameter(typ).into());
+        let parameter_value = Self::map_type(parameter_type, |typ| {
+            let value = self.builder.add_parameter(typ);
+            if mutable {
+                self.new_mutable_variable(value)
+            } else {
+                value.into()
+            }
+        });
 
         self.definitions.insert(parameter_id, parameter_value);
     }
 
+    /// Create a new mutable storage slot and store into it the given initial value of the variable.
```

Comment:
nit: It should be understandable from the context, though if we can use a different term than "storage slot" that would be great, due to contracts having the concept of a "storage slot"